app/controllers/collections_controller.rb

```diff
@@ -189,4 +264,99 @@ def respond_with_error(message, status_code)
     end
   end
 
+  # Uploads a document given as json content or responds with an error if appropriate
+  def upload_document_using_json(filename, json_content, collection_name)
+    if File.exists? filename
+      raise ResponseError.new(412), "The file \"#{File.basename filename}\" has already been uploaded to the collection #{collection_name}"
+    else
+      Rails.logger.debug("Writing uploaded document contents to new file #{filename}")
+      create_file(filename, json_content)
+      filename
+    end
+  end
+
+  # Uploads a document given as a http multipart uploaded file or responds with an error if appropriate
+  def upload_document_using_multipart(absolute_filename, file, collection_name)
+    filename = File.basename absolute_filename
+    if !file.is_a? ActionDispatch::Http::UploadedFile
+      raise ResponseError.new(412), "Error in file parameter."
+    elsif file.blank? or file.size == 0
+      raise ResponseError.new(412), "Uploaded file \"#{filename}\" is not present or empty."
+    elsif File.exists? absolute_filename
+      raise ResponseError.new(412), "The file \"#{filename}\" has already been uploaded to the collection #{collection_name}"
+    else
+      Rails.logger.debug("Copying uploaded document file from #{file.tempfile} to #{absolute_filename}")
+      FileUtils.cp file.tempfile, absolute_filename
+      absolute_filename
+    end
+  end
+
+  # Updates the source of a document in the JSON formatted Item graph
+  def update_document_source_in_graph(json_graph, document_identifier, document_source)
+    json_graph.each do |graph_entry|
+      if graph_entry['dcterms:identifier'] == document_identifier
+        # Escape any filename spaces with '%20' as URIs with spaces are flagged as invalid when RDF loads
+        graph_entry.update({'dcterms:source' => {'@id' => "file://#{document_source.sub(" ", "%20")}"}})
+        json_graph
+      end
+    end
+  end
+
+  # Validates the request on the add items api call
+  def validate_add_items_request(params)
+    collection = Collection.find_by_name(params[:id])
+    if params[:items].blank?
+      raise ResponseError.new(400), "JSON-LD formatted item metadata must be sent to the api call as a POST request"
```

Comment:
Don't really need the POST request warning stuff since it is limited to a POST request anyway. It doesn't really make sense to send data via a GET request anyway.